lock-api/src/test/java/com/palantir/lock/watch/LockWatchEventTrackerImplTest.java

```diff
@@ -0,0 +1,203 @@
+
+public class LockWatchEventTrackerImplTest {
+    private static final String TABLE_NAME = "test.table";
+    private static final byte[] ROW = new byte[] { 'r', 'o', 'w' };
+    private static final byte[] COL = new byte[] { 'c', 'o', 'l' };
+    private static final byte[] LOL = new byte[] { 'l', 'o', 'l' };
+    private static final LockDescriptor DESCRIPTOR = AtlasCellLockDescriptor.of(TABLE_NAME, ROW, COL);
+    private static final LockDescriptor OTHER_DESCRIPTOR = AtlasCellLockDescriptor.of(TABLE_NAME, ROW, LOL);
+    private static final LockWatchRequest ROW_PREFIX_REQ = LockWatchRequest.of(
+            ImmutableSet.of(LockWatchReferences.rowPrefix(TABLE_NAME, ROW)));
+
+    private final LockWatchEventTracker tracker = new LockWatchEventTrackerImpl();
+
+    @Test
+    public void stateWithNoUpdatesIsNotWatched() {
+        VersionedLockWatchState state = tracker.currentState();
+        assertThat(state.version()).isEmpty();
+        assertThat(state.lockWatchState(DESCRIPTOR)).isEqualTo(LockWatchInfo.NOT_WATCHED);
+    }
+
+    @Test
+    public void lockWatchCreatedWithoutOpenLockGivesNoInformationOnLocks() {
+        UUID leaderId = UUID.randomUUID();
+        LockWatchStateUpdate update = LockWatchStateUpdate.update(leaderId, 0, ImmutableList.of(
+                LockWatchCreatedEvent.builder(ROW_PREFIX_REQ, UUID.randomUUID()).build(0)));
+
+        tracker.updateState(update);
+        VersionedLockWatchState state = tracker.currentState();
+        assertThat(state.version()).isEqualTo(OptionalLong.of(0));
+        assertThat(state.leaderId()).isEqualTo(leaderId);
+        assertThat(state.lockWatchState(DESCRIPTOR)).isEqualTo(LockWatchInfo.NOT_WATCHED);
+    }
+
+    @Test
+    public void lockWatchCreatedAfterOpenLockMeansLockIsLocked() {
+        UUID leaderId = UUID.randomUUID();
+        UUID requestId = UUID.randomUUID();
+        LockWatchStateUpdate update = LockWatchStateUpdate.update(leaderId, 1, ImmutableList.of(
+                LockWatchOpenLocksEvent.builder(ImmutableSet.of(DESCRIPTOR), requestId).build(0),
+                LockWatchCreatedEvent.builder(ROW_PREFIX_REQ, requestId).build(1)));
+
+        tracker.updateState(update);
+        VersionedLockWatchState state = tracker.currentState();
+        assertThat(state.version()).isEqualTo(OptionalLong.of(1));
+        assertThat(state.leaderId()).isEqualTo(leaderId);
+        assertThat(state.lockWatchState(DESCRIPTOR)).isEqualTo(LockWatchInfo.of(LOCKED, OptionalLong.of(0)));
+        assertThat(state.lockWatchState(OTHER_DESCRIPTOR)).isEqualTo(LockWatchInfo.of(UNLOCKED, OptionalLong.empty()));
+    }
+
+    @Test
+    public void lockCanBeUpdated() {
+        UUID leaderId = UUID.randomUUID();
+        LockWatchStateUpdate update = LockWatchStateUpdate.update(leaderId, 1, ImmutableList.of(
+                        LockEvent.builder(ImmutableSet.of(DESCRIPTOR)).build(0),
+                        UnlockEvent.builder(ImmutableSet.of(DESCRIPTOR)).build(1)));
```

Comment:
The assumption here, I guess, is that these events are already filtered to be relevant to things that are being watched - here this would mean that the registration of some watch covering `DESCRIPTOR` has already taken place in the past.